### PR TITLE
Add new flash attn features to cuDNN SDPA API and remove fused attn

### PR DESCRIPTION
**Pull request by Cjkkkk**
* add variable sequence length: Accepts two additional tensor seqlen_q and seqlen_kv to indicate the non padded length to reduce computation.
* add MQA/GQA.
* add broadcast bias: bias can be broadcast on batch/head dim.
* add dbias calculation.
* remove fused attn and default to flash attn.

### CODE REVIEW

**Comment by Cjkkkk** [comment]
@superbobry Hi Sergei, could you help review this PR?

**Comment by Cjkkkk** [comment]
@superbobry hi Sergei, any updates on this?

**Comment by superbobry** [comment]
No updates just yet, sorry. I will review some time tomorrow.

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 1]
```diff
@@ -41,10 +41,42 @@ class AttentionLayout(Enum):
...
   BNTH = 1
 
+class MaskType(Enum):
+  NO_MASK = 0
+  PADDING = 1
+  CAUSAL = 2
+  PADDING_CAUSAL = 3
+  ALIBI = 4
+
+def convert_mask_type_to_string(mask_type: MaskType) -> str:
```
Comment (superbobry) [inline jax/_src/cudnn/fused_attention_stablehlo.py:51]: I think this is identical to `mask_type.name`.

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 2]
```diff
@@ -41,10 +41,42 @@ class AttentionLayout(Enum):
...
+  elif mask_type == MaskType.CAUSAL:
+    return "CAUSAL"
+  elif mask_type == MaskType.PADDING_CAUSAL:
+    return "PADDING_CAUSAL"
+  elif mask_type == MaskType.ALIBI:
+    return "ALIBI"
+  else:
+    raise ValueError(f"Unexpected mask type: {mask_type}")
+
+def has_padding(mask_type: MaskType) -> bool:
```
Comment (superbobry) [inline jax/_src/cudnn/fused_attention_stablehlo.py:65]: Make this a property of `MaskType`.

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 3]
```diff
@@ -169,157 +200,171 @@ def create_dot_product_attention_backend_config(batch,
...
 def check_cudnn_version():
   # check if cuDNN is installed
   if cuda_versions is None:
     raise RuntimeError("cuDNN is not detected.")
   return cuda_versions.cudnn_get_version()
 
+def check_compute_capability(cc):
+  if cuda_versions is None:
+    raise RuntimeError("cuDNN is not detected.")
+  for i in range(jax.device_count()):
```
Comment (superbobry) [inline jax/_src/cudnn/fused_attention_stablehlo.py:305]: FYI GPU devices now have `.compute_capability` as an attribute, see e.g. https://github.com/google/jax/blob/210f8bbfcac50aa89a7d97fc9b59e673d3f3e626/jax/_src/test_util.py#L424.

File: tests/fused_attention_stablehlo_test.py [hunk 4]
```diff
@@ -107,55 +154,50 @@ def sdpa_train_ref(query: Array,
...
-    #  even when it passes: https://github.com/google/jax/issues/20438
-    self.skipTest('Failing with various errors')
-    return
-
+    super().setUp()
     if jax.device_count() < 4:
       self.skipTest("Requires more than 4 devices.")
     try:
       cudnn_version = check_cudnn_version()
+      check_compute_capability((80, 90))
```
Comment (superbobry) [inline tests/fused_attention_stablehlo_test.py:175]: FYI we have `jtu.is_cuda_compute_capability_at_least` which does this.

File: tests/fused_attention_stablehlo_test.py [hunk 5]
```diff
@@ -107,55 +154,50 @@ def sdpa_train_ref(query: Array,
...
   @jtu.sample_product(
       batch_size=[4],
-      seq_len=[256, 1024],
+      seq_len=[1024],
       num_heads=[8],
       head_dim=[64, 128],
       use_bias=[False, True],
-      use_mask=[False, True],
-      is_causal_mask=[False],
+      mask_type=[MaskType.NO_MASK],
```
Comment (superbobry) [inline tests/fused_attention_stablehlo_test.py:188]: Does it make sense to test with different masks here?

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 6]
```diff
@@ -41,10 +41,42 @@ class AttentionLayout(Enum):
   BTNH = 0
   BNTH = 1
 
+class MaskType(Enum):
+  NO_MASK = 0
```
Comment (superbobry) [inline jax/_src/cudnn/fused_attention_stablehlo.py:45]: OOC why not use `None` instead, when a mask is not specified?

**Review by superbobry — APPROVED** [review]
I did my best to read through, but these large diffs are really hard to get through. Please send smaller PRs for any follow up changes.

I would also recommend to ask someone from NVidia to review for CUDNN APIs etc.

**Comment by Cjkkkk** [comment]
> I did my best to read through, but these large diffs are really hard to get through. Please send smaller PRs for any follow up changes.
> 
> I would also recommend to ask someone from NVidia to review for CUDNN APIs etc.

Understood, sorry for the large PR, i will create smaller one next time. I think people from Nvidia don't have access to approve and merge the PR?

**Comment by superbobry** [comment]
Please address the comments, and the we can merge.

File: tests/fused_attention_stablehlo_test.py (hunk 4 above)
Comment (Cjkkkk) [reply tests/fused_attention_stablehlo_test.py:175]: cudnn flash attn only support 8.0, 9.0 not 8.6. So at least is not enough.

File: jax/_src/cudnn/fused_attention_stablehlo.py (hunk 1 above)
Comment (Cjkkkk) [reply jax/_src/cudnn/fused_attention_stablehlo.py:51]: fixed

File: tests/fused_attention_stablehlo_test.py (hunk 5 above)
Comment (Cjkkkk) [reply tests/fused_attention_stablehlo_test.py:188]: Added causal. padding is tested in `test_sdpa_var_seq`

**Comment by Cjkkkk** [comment]
> Please address the comments, and the we can merge.

Comments addressed, sorry about the delay.

File: tests/fused_attention_stablehlo_test.py (hunk 4 above)
Comment (superbobry) [reply tests/fused_attention_stablehlo_test.py]: Sure, my point was that you can use a `jtu` API to do the same check.

**Comment by superbobry** [comment]
Can you squash the PR please?

File: jax/_src/cudnn/fused_attention_stablehlo.py (hunk 6 above)
Comment (Cjkkkk) [reply jax/_src/cudnn/fused_attention_stablehlo.py:45]: Just a choice to make it more explicit
